packages/react-router/README.md

```diff
@@ -0,0 +1,378 @@
+
+```tsx
+// Example in Koa...
+
+const app = new Koa();
+app.use((ctx) => {
+  ctx.body = renderToString(
+    <Router url={ctx.URL}>
+      <App />
+    </Router>,
+  );
+});
+```
+
+This component basically just composes the `Prefetcher` component, serialization for the `NoMatch` component, alongside a number of context providers (`RouterContext`, `CurrentUrlContext`). If you wish to omit some features from your final bundle, you can skip the `Router` component and render those primitives instead.
+
+### `<Route />`
+
+Interacts with the `Router` instance to register behavior when the location matches a URL. There is only one mandatory prop, `match`, which specifies the URL to match. This prop can be one of the following:
+
+- a `string`, in which case it will match any URL with a pathname that exactly matches, or that exactly matches except for a trailing `/`,
+- a `RegExp`, in which case it will match any URL with a pathname that matches the regex, or,
+- a function that takes an [`EnhancedURL` object](#enhancedurl), and returns a boolean indicating whether or not to match.
+
+A `<Route />` can then specify the behavior to use for the matching URLs by specifying any set of the following props:
+
+- `render`: a function that is called with an `EnhancedURL` object, and returns JSX to use when the route matches.
+- `renderPrefetch`: a function that is called with a `URL` object, and returns JSX to render when the route is [prefetched](#prefetcher).
+- `redirect`: any value accepted by the [`Redirect` component’s `to` prop](#redirect), which will be used to render a `Redirect` component when the route matches (should not be provided if `render` is also specified).
+
+The following example shows a component created by [`@shopify/react-async`](https://github.com/Shopify/quilt/tree/master/packages/react-async) that is rendered and prefetched when the current path is exactly `/products` or `/products/`:
+
+```tsx
+import React from 'react';
+import {Route} from '@shopify/react-router';
+import {createAsyncComponent} from '@shopify/react-async';
+
+const Products = createAsyncComponent({
+  load: () => import('./Products'),
+});
+
+export function ProductsRoute() {
+  return (
+    <Route
+      match="/products"
+      render={() => <Products />}
+      renderPrefetch={() => <Products.Prefetch />}
+    />
+  );
+}
+```
+
+This example shows how to extract a part of the path from the URL that your component needs to know (in this case, an ID in the form `/products/ID`):
+
+```tsx
+import React from 'react';
+import {Route} from '@shopify/react-router';
+import Product from './Product';
+
+const PRODUCT_MATCH = /^\/products\/(\d+)(\/|$)/;
```

Comment:
I get the intention of using more language primitives instead of arcane weakly typed DSLs, but I think a static url with one or more dynamic segments is probably common enough that a specific abstraction would be preferable than a bunch of boilerplate. Maybe something like

```
  <DynamicSegmentRoute
     match="/product/:id" 
     render{(url, segments) => {
        const [id] = segments;
        return <Product id={id} />;
     }}
  />
```

Reply:
Or maybe it's built into the basic `Route` /shrug

Reply:
➕To all of the above. It would make a refactor from `react-router` to this much gentler.

Reply:
Hmm, I am definitely open to this idea, but my feeling is that it's better in this case for the library to be the primitive, and consuming projects can explore ideas that work better for them personally if they care. I specifically designed this library (`NoMatch` in particular) to work well with composition (unlike react-router, where `Switch`s break composition).

If you wanted the API Mat suggested, you could do it super easily in userspace:

```tsx
function DynamicSegmentRoute({match, render}, {match: string, render: (url: URL, segments: string[]) => React.ReactNode}) {
  return <Route match={toRegex(match)} render={(url) => render(url, getDynamicSegments(match, url)} />
}
```

If you wanted the exact React Router PR (even just for rolling it out incrementally), you can get that too:

```tsx
function DynamicSegmentRoute({path, render}: ReactRouterRouteProps) {
  return <Route match={toRegex(path)} render={(url) => render(url, getParamsObject(match, url)} />
}
```

Reply:
Don't you think that we'll always use `toRegex()` and because of that it's better to have it in the library itself?
I mean, it's a lot easier to read and understand what `/product/:id` does than `/^\/products\/(\d+)(\/|$)/`

Reply:
As soon as you have the idea of placeholders you are back in the land of being unable to provide any type guarantees about those extracted bits.  You will not always end up using a regency - many routes across our apps do not have dynamic pieces to them. The above were just for illustrating that you can use the regex/ function form here to create any API you like.

I don’t disagree that the regex is more awkward, but it also supports the full power of JS, rather than the confusing subset allowed by the path to regex stuff (https://github.com/pillarjs/path-to-regexp)

Reply:
I don't think there's an issue with allowing pure regexes for cases where we need them, I just feel like the path notation is a common enough convention that there's not that good of a reason to not support it. It's only less type safe than a regex if we pass it down as a dictionary. As an array of strings it's not actually any different than the value returned from regex matching.

It's likely that we'll end up using a tiny wrapper route component everywhere as is, which is maybe fine but begs the question of why not just support the common usecase directly.

Reply:
How would people feel if `render` got called with two arguments: the current URL and an array of strings, which in the case of a string `match` would be any segment matching an `:foo` pattern (restricted to just the simplest case), and in the case of a regular expression would be the capture groups?